### PR TITLE
NAS-141173 / jbof: shelf-attach workarounds (hardwire_host retry + drive-slot recovery)

### PR DESCRIPTION
**Pull request by darkfiberiru**
## Summary

Two middleware-only workarounds for corner cases that surface during
JBOF shelf attach. Each addresses a distinct firmware issue we've
reproduced and escalated upstream; the underlying fixes will land in
a future firmware release. Until then these workarounds let
`jbof.create` succeed in the affected scenarios instead of returning
errors or silently leaving drives unattached.

Both are middleware-only — no kernel changes, no schema changes, no
on-disk format changes. They run inside `jbof.create` / `jbof.update`'s
`hardwire_dataplane` and are no-ops on hosts/shelves that don't hit
the corner cases.

Tracked in NAS-141173.

## Commits

- jbof: retry hardwire_host with fabric-card reset on failure
- jbof: recover drive slots missing from fabric discovery

## Test plan

- [x] Affected shelf, single-controller: 60+ add/remove iterations,
      both patches firing as needed, full attach every iter
- [x] Affected shelf, HA: 20/20 add/remove iterations at 120 s gap,
      full attach on both controllers every iter, both patches
      exercised
- [x] Unaffected shelf, single-controller: 20+ iterations, neither
      patch fires, no performance regression vs pre-patch baseline,
      no new errors on shelf BMC
- [x] HA failure modes: rebooted each controller in turn (BACKUP +
      MASTER w/ failover), IPMI hard power-cycle one controller —
      shelf state persists, attaches restored automatically
- [x] Host reboot: both shelves auto-reattach from DB on
      configure_jbof, all drives reconnect without intervention

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/middlewared/middlewared/plugins/jbof/crud.py [hunk 1]
```diff
@@ -644,6 +665,193 @@ async def hardwire_host(self, mgmt_ip, shelf_index, schema, verrors):
...
+        if not expected:
+            return []
+        missing = set()
+        for ip in jbof_ips:
+            try:
+                proc = await asyncio.create_subprocess_exec(
+                    'nvme', 'discover', '-t', 'rdma', '-a', ip, '-s', '4420',
+                    stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE,
+                )
+                out, _ = await proc.communicate()
```
Comment (bmeagherix) [inline src/middlewared/middlewared/plugins/jbof/crud.py]: Can this hang?

Would it make sense to wrap the `communicate` inside a `asyncio.wait_for` with a sensible timeout?

File: src/middlewared/middlewared/plugins/jbof/crud.py [hunk 2]
```diff
@@ -644,6 +665,193 @@ async def hardwire_host(self, mgmt_ip, shelf_index, schema, verrors):
...
+
+        Blocks until the BMC reports the restart has completed (~28s). Default
+        Redfish client timeout is 10s, so an explicit 90s timeout is used here.
+        Returns True on success, False on failure.
+        """
+        uri = f'/redfish/v1/Chassis/IOM{iom_num}/NetworkAdapters/1/Actions/Oem/VikingEnterpriseSolutions.Reset'
+        try:
+            RedfishClient.cache_get(mgmt_ip).post(
+                uri, data={'ResetType': 'GracefulRestart'}, timeout=90,
+            )
```
Comment (bmeagherix) [inline src/middlewared/middlewared/plugins/jbof/crud.py]: Check the return value from the post?

File: src/middlewared/middlewared/plugins/jbof/crud.py [hunk 3]
```diff
@@ -644,6 +665,193 @@ async def hardwire_host(self, mgmt_ip, shelf_index, schema, verrors):
...
+
+        reset_type must be 'ForceOff' or 'On'. The BMC serializes ForceOff
+        requests at ~2.3s/slot regardless of issue order, so an explicit
+        timeout is required.
+        """
+        uri = f'/redfish/v1/Chassis/2U24/Drives/{slot_id}/Actions/Drive.Reset'
+        try:
+            RedfishClient.cache_get(mgmt_ip).post(
+                uri, data={'ResetType': reset_type}, timeout=60,
+            )
```
Comment (bmeagherix) [inline src/middlewared/middlewared/plugins/jbof/crud.py]: Check the return value from the post?

**Comment by darkfiberiru** [comment]
@bmeagherix All 3 should be addressed

**Review by bmeagherix — APPROVED** [review]
